### ISSUE TITLE
Xonsh environment returns wrong value for undefined variables and also missing some OS environment variables.

### ISSUE THREAD

**Issue by bobhy** [xonsh/xonsh#3739]
Related to #3703 and PR #3377, but with more feeling.

## Steps to Reproduce / Current Behavior

```
# Case 1: not defined in OS environ, but Xonsh has a known default
$ os.environ['ANSICON']
KeyError: 'ANSICON'
$ ls $ANSICON
ls: cannot access '': No such file or directory
# As expected (but only because ANSICON has defined default of ``).

# Case 2: not defined in OS environ or Xonsh
$ ls $NO_SUCH_VARIABLE
ls: cannot access '$NO_SUCH_VARIABLE': No such file or directory
# Expected empty string.

# Case 3: defined in Xonsh environ (and synced to OS environ)
$ ls $SHELL_TYPE
ls: cannot access 'prompt_toolkit': No such file or directory
# As expected

# Case 4: defined in OS environment, but not in Xonsh environment
$ os.environ["WINDIR"]
'C:\\WINDOWS'
$ ls $WINDIR
ls: cannot access '$WINDIR': No such file or directory
# expected 'C:\WINDOWS'.

# Case 5: not defined in OS environment but not a Xonsh configuration parameter
$ "ANSICON" in os.environ
False
$ "ANSICON" in __xonsh__.env
True
# Expected 2nd to be False if first is False (issue #3703)
```

## Expected Behavior

3 bugs here:
1. OS environment variables that don't happen to match a key in `xonsh.environ.DEFAULT_VARS` are not present in `__xonsh__.env` and `$<env_var>`.  (Case 4 but also Case 2)
2. Reference to an undefined (xonsh) environment variable should evaluate to empty string at least in subproc mode. (Case 2 but also Case 4 and Case 1).
3. Xonsh environment returns a value to python code when the OS environment value is not defined. (case 5, issue of #3703).  This might be acceptable for variables that xonsh defines and consumes, but is clearly a lie for environment variables that other programs are supposed to define for xonsh.

We can address case 4 by simply ensuring *all* the OS environment variables are loaded into Xonsh at startup.

We might be able to fix the rest of the cases (and resolve #3703)  if we implement an is-defined function so code that cares about it can tell whether it was really defined.  And we'd return the defined default for references to undefined environment variables that had them and hard code empty string for other undefined variables..  For example:

``` 
$ __xonsh__.env["ANSICON"]
''
$ "ANSICON" in __xonsh__.env
False

$ __xonsh__.env["NO_SUCH_VARIABLE"]
''
$ "NO_SUCH_VARIABLE" in __xonsh__.env
False
```

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by laloch** [comment]
Hi @bobhy, thanks for opening the issue.

> OS environment variables that don't happen to match a key in xonsh.environ.DEFAULT_VARS are not present in __xonsh__.env and $<env_var>. (Case 4 but also Case 2)

For me, the above statement is clearly false. Also the code below seems to prove this is wrong.
https://github.com/xonsh/xonsh/blob/5aed9ee5cab85039a5863846b4dd9d2059d824a5/xonsh/environ.py#L1726-L1737

> Reference to an undefined (xonsh) environment variable should evaluate to empty string at least in subproc mode. (Case 2 but also Case 4 and Case 1).

That's a matter of opinion. I prefer always raising `KeyError`. Accessing undefined variables is in my opinion always a bug regardless of the execution mode.

> Xonsh environment returns a value to python code when the OS environment value is not defined. (case 5, issue of #3703). This might be acceptable for variables that xonsh defines and consumes, but is clearly a lie for environment variables that other programs are supposed to define for xonsh.

I don't understand. `$ANSICON` is a variable with default defined by Xonsh:
```
$ ${...}.help("ANSICON")
ANSICON:

This is used on Windows to set the title, if available.

default: ''
configurable: False
```

**Comment by laloch** [comment]
@bobhy, do you know about `$UPDATE_OS_ENVIRON` switch?

**Comment by laloch** [comment]
Ok, I got confused by this:
> We can address case 4 by simply ensuring all the OS environment variables are loaded into Xonsh at startup.

But that's exactly what Xonsh does now:
https://github.com/xonsh/xonsh/blob/5aed9ee5cab85039a5863846b4dd9d2059d824a5/xonsh/environ.py#L1726-L1737

**Comment by bobhy** [comment]
You're right, I cannot repro the missing WINDIR now.  It has the expected value when I run the test now.  WTF?

But the larger issue remains: user has every right to expect that reference to undefined environment in subproc mode should evaluate to empty string, not raise KeyError.  That is the point of setting the empty string as default value for known environment vars in the first place, and it ought to be extended to unknown environ vars as well. 

In program mode (and in xonsh code itself), you certainly must be able to test whether an env var is in the current environment, but you don't *need* the KeyError, since all the places that reference env use `.get(<key>...)` or similar defense against it.  Since this code is under our control, we change all the references that matter and can document whatever rules we come up with.

I agree it's a bit unexpected to return empty string in prog mode for undefined vars.  The advantage of my proposal is that it does not require knowing whether you're in program or subproc mode (I'd need serious help to figure out how to do this!).  It's also the case that most (all?) existing code that tests for existence of the variable already uses `in`.  I think we can document this for future developers.

Do you see any fatal flaws?  I'm working on a PR for this, and will have that available for review today/tomorrow.

**Comment by laloch** [comment]
@scopatz, what's your opinion. I still think accessing undefined variables should raise `NameError` (not `KeyError`) regardless of the execution mode. We are Python after all. I don't like the new behavior of empty default environment variables either. It's very non-pythonic.

**Comment by bobhy** [comment]
I'm open to pointers to the place to implement the empty string default only for subproc mode, if that would ease your discomfort.  I agree that doing it in python mode is a (managable?) stretch.

But the convention is well established in other shells  for shell scripting and command line hacking. Here's BASH handling 2 undefined environment variables:
```
$ echo wibble wabble $TERM "frotz${ANSICON}stuff" "king${NO_SUCH_VARIABLE}fink"
wibble wabble xterm-256color frotzstuff kingfink
```

**Comment by bobhy** [comment]
Updated title to narrow scope. I can no longer repro Case 4 above, the problem of OS environment variables not being present in xonsh env!

**Comment by bobhy** [comment]
Narrowing the scope of this problem some more.  
Now, I'll say that there are 2 issues:
```
# Case 5: not defined in OS environment but not a Xonsh configuration parameter
@ "ANSICON" in os.environ
False
@ "ANSICON" in __xonsh__.env
True
# Expected 2nd to be False if first is False (issue #3703)

# Case 5a: not in OS environment, access via .get() 
@ __xonsh__.env.get("ANSICON", "foo")
""
# Expected value to be "foo"
```

This is not a problem for Xonsh's own configuration parameters (stored in environment), Xonsh code can reasonably be expected to know the registered default that will be returned.  But I'll maintain that it is a problem for TERM, VIRTUAL_ENV and ANSICON (and OLDPWD?)  These are defined by external software and really might be undefined, so code really ought to be able to rely on MutableMapping interface rules.

**Comment by anki-code** [comment]
After reading and testing all cases I prefer to say that they are fixed.
The case `echo $UNKNOWN` is working as expected.
If certain case requires deep dive and fix please open issue with only this case.
